sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/Analyzer.scala

```diff
@@ -77,6 +77,7 @@ class Analyzer(
       ExtractWindowExpressions ::
       GlobalAggregates ::
       UnresolvedHavingClauseAttributes ::
+      RemoveEvaluationFromSort ::
```

Comment:
The most optimal way is we have a perfect cost model that can predict what we are trading off (network vs cpu). Minus that, I think just always projecting is the approach that makes more sense is most common cases, because:

1. It is hard to quantify the difference.
2. I/O (network, disk) is rarely the bottleneck here, especially with more SSDs and 10Gbps network.
3. Most of the time order by is just ordering by a field, and this won't hurt that case.
4. If there is a complex expression, doing the eval many times during sorting is bad.

The alternative, which is probably even better, is for the sorter itself to always project out the sort key. It might make more sense there, but is slightly more complicated to write I think.